---
title: "Single Cell Tumor Host Interaction (scTHI)"
author: "Author"
date: "April, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

***

### Introduction
The detection of tumor-host interaction is a major challange in cancer understanding. Recent advantages in single cell next generation sequencing have provided new insights in unveiling  tumor-microenvironment cells comunications. 
The scTHI package provides an rank-based approach to test the expression in single cell data of a set of interacting ligands and receptors avaiable in literature [1]. 

### Quick start

```{r,message=FALSE,warning=FALSE, eval=FALSE, echo=TRUE}
library(devtools)
install_github("miccec/scTHI")
library(scTHI)
```

### Input Data
As input, the scTHI package expects a matrix of count data (or normalized counts) obtained from single cell RNA-seq experiment, where rows are genes presented with Hugo Symbols and columns are cells. For a practical demonstration, let’s use the scRNA-seq data available on the Broad Institute Single-Cell Portal (https://portals.broadinstitute.org/single_cell/study/single-cell-analysis-in-pediatric-midline-gliomas-withhistone-h3k27m-mutation) [2]. The dataset provides 3,321 scRNA-seq profiles from six primary Glioma (H3K27M-glioma), and includes both malignant cells and several tumor microenvironment cell types, as immune cells and oligodendrocytes. Our goal is to identify the significant ligand-receptor interactions that are established between cancer and immune cells present in the tumor microenvironment. For this reason, after downloading H3K27M-glioma data we preprocessed them, selecting only one sample of interest (i.e. BCH836), removing zero genes, trasforming data in TMP e log2, and appling quantile normalization. Processed data can be loaded as follows:

```{r,message=FALSE,warning=FALSE, eval=FALSE, echo=TRUE}
load(system.file( "H3K27_PatientBCH836.RData", package = "scTHI", mustWork = TRUE))
```

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=FALSE}
#load("H3K27_PatientBCH836.RData")
```

H3K27.meta includes the annotation of each cell, so as to identify the tumor cells and the immune cells in BCH836 patient:

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=TRUE}
table(H3K27.meta$Type)
```

```{r,message=FALSE,warning=FALSE, eval=FALSE, echo=TRUE}
Malignant <- rownames(H3K27.meta)[H3K27.meta$Type == "Malignant"]
Immune <- rownames(H3K27.meta)[H3K27.meta$Type == "Immune cell"]
```

### Run scTHI
Let's run `scTHI` in order to detect significant ligand-receptor interactions between cancer and immune cell. P-value is estimated by permuting the mates of interaction pairs. 
Important to consider is that interactions are not symmetric, so partnerA expression is considered on the first cluster (in the example the Malignant cells), and partnerB expression is considered on the second cluster (i.e. Immune cells). 

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=TRUE}
result <- scTHI(expMat = H3K27, cellCusterA = Malignant, cellCusterAName = "Malignant", cellCusterB = Immune, 
                cellCusterBName = "Immune", topRank = 15, fileNameBase = "PatientBCH836", filterCutoff = 0.50, 
                foldChange = FALSE, useData = FALSE, PValue = TRUE, nPermu = 100, ncore = 48)
```

As result are shown only significant interaction pairs, for which partner A and partner B are expressed in at least 50% of the cells of the respective clusters, as set in *filterCutoff* argument. The scTHI score indicates the percentage of cells that express the interaction pair at the top of the ordered gene list. In this regard, we consider a gene expressed in a cell, if it is in the top 15% of the ranked gene list, as set in *topRank* argument. Result also includes two columns with the average expression values of each partner in the respective clusters.

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=FALSE}
#load("PatientBCH836_MalignantVSImmune.RData")
```

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=TRUE}
head(result)
```

The `scTHI` function also allows the user to calculate a fold change of expression of each partner with respect to a third cluster of cells of interest. Suppose we want to check whether the two partners of each interaction are more or less expressed than the oligodendrocyte cells. We can run again `scTHI` with the results previously obtained and changing the *foldChange* and *useData* arguments.

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=TRUE}
result <- scTHI(expMat = H3K27, cellCusterA = Malignant, cellCusterAName = "Malignant", cellCusterB = Immune, 
                cellCusterBName = "Immune", topRank = 15, fileNameBase = "PatientBCH836_plusFC", 
                filterCutoff = 0.50, foldChange = TRUE, useData = TRUE, dataToUse = result, 
                cellsToCompare = Oligodendrocyte, cellsToCompareNAME = "Oligodendrocyte", 
                expMatToCompare = H3K27, PValue = FALSE)
```

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=FALSE}
#load("PatientBCH836_plusFC_MalignantVSImmune.RData")
```

```{r,message=FALSE,warning=FALSE, eval=TRUE, echo=TRUE}
head(result)
```


##### References
1. Vento-Tormo, R., Efremova, M., Botting, R. A., Turco, M. Y., Vento-Tormo, M., Meyer, K. B., ... & Gardner, L. (2018). Single-cell reconstruction of the early maternal–fetal interface in humans. Nature, 563(7731), 347.

2. Filbin, M. G., Tirosh, I., Hovestadt, V., Shaw, M. L., Escalante, L. E., Mathewson, N. D., ... & Haberler, C. (2018). Developmental and oncogenic programs in H3K27M gliomas dissected by single-cell RNA-seq. Science, 360(6386), 331-335.

