<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>scTHI vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">scTHI vignette</h1>



<hr />
<div id="introduction" class="section level3">
<h3>Introduction</h3>
<p>The detection of tumor-host interaction is a major challange in cancer understanding. Recent advantages in single cell next generation sequencing have provided new insights in unveiling tumor-microenvironment cells comunications. The scTHI package provides an rank-based approach to test the expression in single cell data of a set of interacting ligands and receptors avaiable in literature [1].</p>
</div>
<div id="quick-start" class="section level3">
<h3>Quick start</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;miccec/scTHI&quot;</span>)
<span class="kw">library</span>(scTHI)</code></pre></div>
</div>
<div id="input-data" class="section level3">
<h3>Input Data</h3>
<p>As input, the scTHI package expects a matrix of count data (or normalized counts) obtained from single cell RNA-seq experiment, where rows are genes presented with Hugo Symbols and columns are cells. For a practical demonstration, let’s use the scRNA-seq data available on the Broad Institute Single-Cell Portal (<a href="https://portals.broadinstitute.org/single_cell/study/single-cell-analysis-in-pediatric-midline-gliomas-withhistone-h3k27m-mutation" class="uri">https://portals.broadinstitute.org/single_cell/study/single-cell-analysis-in-pediatric-midline-gliomas-withhistone-h3k27m-mutation</a>) [2]. The dataset provides 3,321 scRNA-seq profiles from six primary Glioma (H3K27M-glioma), and includes both malignant cells and several tumor microenvironment cell types, as immune cells and oligodendrocytes. Our goal is to identify the significant ligand-receptor interactions that are established between cancer and immune cells present in the tumor microenvironment. For this reason, after downloading H3K27M-glioma data we preprocessed them, selecting only one sample of interest (i.e. BCH836), removing zero genes, trasforming data in TMP e log2, and appling quantile normalization. Processed data can be loaded as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scTHI)
<span class="kw">load</span>(<span class="kw">system.file</span>(<span class="st">&quot;H3K27_PatientBCH836.RData&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;scTHI&quot;</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>))</code></pre></div>
<p>H3K27.meta includes the annotation of each cell, so as to identify the tumor cells and the immune cells in BCH836 patient:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(H3K27.meta<span class="op">$</span>Type)</code></pre></div>
<pre><code>## 
##          Filter     Immune cell       Malignant Oligodendrocyte 
##               2              53             438              34</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Malignant &lt;-<span class="st"> </span><span class="kw">rownames</span>(H3K27.meta)[H3K27.meta<span class="op">$</span>Type <span class="op">==</span><span class="st"> &quot;Malignant&quot;</span>]
Immune &lt;-<span class="st"> </span><span class="kw">rownames</span>(H3K27.meta)[H3K27.meta<span class="op">$</span>Type <span class="op">==</span><span class="st"> &quot;Immune cell&quot;</span>]
Oligodendrocyte &lt;-<span class="st">  </span><span class="kw">rownames</span>(H3K27.meta)[H3K27.meta<span class="op">$</span>Type <span class="op">==</span><span class="st"> &quot;oligodendrocyte&quot;</span>]</code></pre></div>
</div>
<div id="run-scthi" class="section level3">
<h3>Run scTHI</h3>
<p>Let’s run <code>scTHI</code> in order to detect significant ligand-receptor interactions between cancer and immune cell. P-value is estimated by permuting the mates of interaction pairs. Important to consider is that interactions are not symmetric, so partnerA expression is considered on the first cluster (in the example the Malignant cells), and partnerB expression is considered on the second cluster (i.e. Immune cells).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">result &lt;-<span class="st"> </span><span class="kw">scTHI</span>(<span class="dt">expMat =</span> H3K27, <span class="dt">cellCusterA =</span> Malignant, <span class="dt">cellCusterAName =</span> <span class="st">&quot;Malignant&quot;</span>, <span class="dt">cellCusterB =</span> Immune, 
                <span class="dt">cellCusterBName =</span> <span class="st">&quot;Immune&quot;</span>, <span class="dt">topRank =</span> <span class="dv">15</span>, <span class="dt">fileNameBase =</span> <span class="st">&quot;PatientBCH836&quot;</span>, <span class="dt">filterCutoff =</span> <span class="fl">0.50</span>, 
                <span class="dt">foldChange =</span> <span class="ot">FALSE</span>, <span class="dt">useData =</span> <span class="ot">FALSE</span>, <span class="dt">PValue =</span> <span class="ot">TRUE</span>, <span class="dt">nPermu =</span> <span class="dv">100</span>, <span class="dt">ncore =</span> <span class="dv">48</span>)</code></pre></div>
<p>As result are shown only significant interaction pairs, for which partner A and partner B are expressed in at least 50% of the cells of the respective clusters, as set in <em>filterCutoff</em> argument. The scTHI score indicates the percentage of cells that express the interaction pair at the top of the ordered gene list. In this regard, we consider a gene expressed in a cell, if it is in the top 15% of the ranked gene list, as set in <em>topRank</em> argument. Result also includes two columns with the average expression values of each partner in the respective clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(result)</code></pre></div>
<pre><code>##                      interationPair interactionType partnerA    partnerB
## PTN_PLXNB2               PTN_PLXNB2          simple      PTN      PLXNB2
## THY1_aXb2 complex THY1_aXb2 complex         complex     THY1 ITGAX:ITGB2
## THY1_aMb2 complex THY1_aMb2 complex         complex     THY1 ITGB2:ITGAM
## APP_CD74                   APP_CD74          simple      APP        CD74
## KLRC2_HLA-E_           KLRC2_HLA-E_          simple    KLRC2       HLA-E
## LRP1_PDGFB               LRP1_PDGFB          simple     LRP1       PDGFB
##                   rnkPartnerA_Malignant expValueA_Malignant
## PTN_PLXNB2                         0.99                9.64
## THY1_aXb2 complex                  0.85                6.94
## THY1_aMb2 complex                  0.85                6.94
## APP_CD74                           0.59                5.02
## KLRC2_HLA-E_                       0.59                4.98
## LRP1_PDGFB                         0.81                6.16
##                   rnkPartnerB_Immune expValueB_Immune SCORE SCOREpValue
## PTN_PLXNB2                      0.83             6.23 0.910        0.03
## THY1_aXb2 complex               0.96      8.08 # 9.18 0.905        0.03
## THY1_aMb2 complex               0.91      9.18 # 6.85 0.880        0.03
## APP_CD74                        1.00            12.97 0.795        0.03
## KLRC2_HLA-E_                    1.00            10.49 0.795        0.03
## LRP1_PDGFB                      0.64             5.11 0.725        0.03</code></pre>
<p>The <code>scTHI</code> function also allows the user to calculate a fold change of expression of each partner with respect to a third cluster of cells of interest. Suppose we want to check whether the two partners of each interaction are more or less expressed than the oligodendrocyte cells. We can run again <code>scTHI</code> with the results previously obtained and changing the <em>foldChange</em> and <em>useData</em> arguments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">result &lt;-<span class="st"> </span><span class="kw">scTHI</span>(<span class="dt">expMat =</span> H3K27, <span class="dt">cellCusterA =</span> Malignant, <span class="dt">cellCusterAName =</span> <span class="st">&quot;Malignant&quot;</span>, <span class="dt">cellCusterB =</span> Immune, 
                <span class="dt">cellCusterBName =</span> <span class="st">&quot;Immune&quot;</span>, <span class="dt">topRank =</span> <span class="dv">15</span>, <span class="dt">fileNameBase =</span> <span class="st">&quot;PatientBCH836_plusFC&quot;</span>, 
                <span class="dt">filterCutoff =</span> <span class="fl">0.50</span>, <span class="dt">foldChange =</span> <span class="ot">TRUE</span>, <span class="dt">useData =</span> <span class="ot">TRUE</span>, <span class="dt">dataToUse =</span> result, 
                <span class="dt">cellsToCompare =</span> Oligodendrocyte, <span class="dt">cellsToCompareNAME =</span> <span class="st">&quot;Oligodendrocyte&quot;</span>, 
                <span class="dt">expMatToCompare =</span> H3K27, <span class="dt">PValue =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(result)</code></pre></div>
<pre><code>##                      interationPair interactionType partnerA    partnerB
## PTN_PLXNB2               PTN_PLXNB2          simple      PTN      PLXNB2
## THY1_aXb2 complex THY1_aXb2 complex         complex     THY1 ITGAX:ITGB2
## THY1_aMb2 complex THY1_aMb2 complex         complex     THY1 ITGB2:ITGAM
## APP_CD74                   APP_CD74          simple      APP        CD74
## KLRC2_HLA-E_           KLRC2_HLA-E_          simple    KLRC2       HLA-E
## LRP1_PDGFB               LRP1_PDGFB          simple     LRP1       PDGFB
##                   rnkPartnerA_Malignant expValueA_Malignant
## PTN_PLXNB2                         0.99                9.64
## THY1_aXb2 complex                  0.85                6.94
## THY1_aMb2 complex                  0.85                6.94
## APP_CD74                           0.59                5.02
## KLRC2_HLA-E_                       0.59                4.98
## LRP1_PDGFB                         0.81                6.16
##                   rnkPartnerB_Immune expValueB_Immune SCORE SCOREpValue
## PTN_PLXNB2                      0.83             6.23 0.910        0.03
## THY1_aXb2 complex               0.96      8.08 # 9.18 0.905        0.03
## THY1_aMb2 complex               0.91      9.18 # 6.85 0.880        0.03
## APP_CD74                        1.00            12.97 0.795        0.03
## KLRC2_HLA-E_                    1.00            10.49 0.795        0.03
## LRP1_PDGFB                      0.64             5.11 0.725        0.03
##                   log2FC_partnerA__MalignantVSOligodendrocyte
## PTN_PLXNB2                                                NaN
## THY1_aXb2 complex                                         NaN
## THY1_aMb2 complex                                         NaN
## APP_CD74                                                  NaN
## KLRC2_HLA-E_                                              NaN
## LRP1_PDGFB                                                NaN
##                   log2FC_partnerB__ImmuneVSOligodendrocyte
## PTN_PLXNB2                                             NaN
## THY1_aXb2 complex                                      NaN
## THY1_aMb2 complex                                      NaN
## APP_CD74                                               NaN
## KLRC2_HLA-E_                                           NaN
## LRP1_PDGFB                                             NaN</code></pre>
<div id="references" class="section level5">
<h5>References</h5>
<ol style="list-style-type: decimal">
<li><p>Vento-Tormo, R., Efremova, M., Botting, R. A., Turco, M. Y., Vento-Tormo, M., Meyer, K. B., … &amp; Gardner, L. (2018). Single-cell reconstruction of the early maternal–fetal interface in humans. Nature, 563(7731), 347.</p></li>
<li><p>Filbin, M. G., Tirosh, I., Hovestadt, V., Shaw, M. L., Escalante, L. E., Mathewson, N. D., … &amp; Haberler, C. (2018). Developmental and oncogenic programs in H3K27M gliomas dissected by single-cell RNA-seq. Science, 360(6386), 331-335.</p></li>
</ol>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
